transpiler/transpiler.go

```diff
@@ -0,0 +1,223 @@
+//
+
+// transpiler convert a subset of Go to Arduino C++ dialect.
+package transpiler
+
+import (
+	"bytes"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"io"
+	"strings"
+)
+
+// Transpile reads Go source code from the given Reader and writes the
+// transpiled Arduino C++ code to the given Writer.
+func Transpile(out io.Writer, in io.Reader, debug io.Writer) error {
+	fset := token.NewFileSet()
+	f, err := parser.ParseFile(fset, "sketch.go", in, 0)
+	if err != nil {
+		return fmt.Errorf("failed to parse file: %v", err)
+	}
+
+	if debug != nil {
+		ast.Fprint(debug, fset, f, nil)
+	}
+
+	for _, d := range f.Decls {
+		if err := handleDecl(out, d); err != nil {
+			return fmt.Errorf("error handling decl %#v: %v", d, err)
+		}
+	}
+	return nil
+}
+
+func handleDecl(out io.Writer, d ast.Decl) error {
+	switch decl := d.(type) {
+	case *ast.GenDecl:
+		return handleGenDecl(out, decl)
+	case *ast.FuncDecl:
+		return handleFuncDecl(out, decl)
+	default:
+		return fmt.Errorf("unsupported decl: %#v", d)
+	}
+}
+
+func handleGenDecl(out io.Writer, gd *ast.GenDecl) error {
+	for _, s := range gd.Specs {
+		vs, ok := s.(*ast.ValueSpec)
+		if !ok {
+			return fmt.Errorf("unsupported spec: %#v", s)
+		}
+		if len(vs.Names) > 1 {
+			return fmt.Errorf("unsupported # of value names: %v", vs.Names)
+		}
+		decl := []string{}
+		if vs.Names[0].Obj.Kind == ast.Con {
+			decl = append(decl, "const")
+		}
+		if len(vs.Values) > 1 {
+			return fmt.Errorf("unsupported # of values: %v", vs.Names)
+		}
+		l, ok := vs.Values[0].(*ast.BasicLit)
+		if !ok {
+			return fmt.Errorf("unsupported value: %#v", vs.Values[0])
+		}
+		if l.Kind != token.INT {
+			return fmt.Errorf("unsupported literal kind: %#v", l.Kind)
+		}
+		decl = append(decl, "int", vs.Names[0].Name, "=", l.Value)
+		fmt.Fprintf(out, "%s;\n", strings.Join(decl, " "))
+	}
+	return nil
+}
+
+func handleFuncDecl(out io.Writer, fd *ast.FuncDecl) error {
+	if fd.Type.Results != nil {
+		return fmt.Errorf("unsupported return type: %#v", fd.Type.Results)
+	}
+	if fd.Type.Params.List != nil {
+		return fmt.Errorf("unsupported param type: %#v", fd.Type.Params)
+	}
+	fmt.Fprintf(out, "void %s() {\n", fd.Name)
+	if err := handleBlockStmt(out, fd.Body); err != nil {
+		return fmt.Errorf("error handling block statement for %q: %v", fd.Name, err)
+	}
+	fmt.Fprintln(out, "}")
+	return nil
+}
+
+func handleBlockStmt(out io.Writer, bs *ast.BlockStmt) error {
+	for _, s := range bs.List {
+		fmt.Fprintf(out, "  ")
+		switch st := s.(type) {
+		case *ast.ExprStmt:
+			if err := handleExpr(out, st.X); err != nil {
+				return fmt.Errorf("error handling expr stmt %v: %v", st.X, err)
+			}
+			fmt.Fprint(out, ";\n")
+		case *ast.AssignStmt:
+			if len(st.Lhs) > 1 {
+				return fmt.Errorf("unsupported # of lhs exprs: %v", st.Lhs)
+
+			}
+			if err := handleExpr(out, st.Lhs[0]); err != nil {
+				return fmt.Errorf("error handling left expr %v: %v", st.Lhs[0], err)
+			}
+			fmt.Fprintf(out, "=")
```

Comment:
`st.Tok` must be checked here; there are [many others](https://gist.github.com/lpereira/e9f0c8866adb89dd9855ad35fc3999e1#file-teste-go-L333-L374) besides simple assignment.